### PR TITLE
Optimize DISTINCT over singleton collections

### PR DESCRIPTION
**Pull request by ranma42**
This does some progress towards fixing #34482, but in several cases the optimization does not trigger because of another intermediate operation.

I will investigate whether it makes sense to combine these operations in a separate PR or if they can be easily handled in this one.

The `Single/First[OrDefault]` case seems to be reasonably tested; for the `Take(1)` case it might be better to add a new test (it is tested, but AFAICT only in combination with `Single/First[OrDefault]`).

### CODE REVIEW

File: src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs [hunk 1]
```diff
@@ -255,6 +255,11 @@ public void ApplyDistinct()
             throw new InvalidOperationException(RelationalStrings.DistinctOnCollectionNotSupported);
         }
 
+        if (Limit is SqlConstantExpression { Value: 0 or 1 })
```
Comment (ranma42) [inline src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs:258]: does it make sense to handle the case of `0`? AFAICT `SELECT` expressions with a `LIMIT 0` could be optimized away heavily (basically only the projection/typing matters; offset, predicate, distinct, ... are all irrelevant)

File: src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs [hunk 2]
```diff
@@ -1901,6 +1906,10 @@ public void ApplyLimit(SqlExpression sqlExpression)
         }
 
         Limit = sqlExpression;
+        if (Offset is null && Limit is SqlConstantExpression { Value: 0 or 1 })
```
Comment (roji) [inline src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs:1909]: nit:

```suggestion

        if (Offset is null && Limit is SqlConstantExpression { Value: 0 or 1 })
```

File: src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs (hunk 1 above)
Comment (roji) [reply src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs:258]: Yeah, I agree - if we want to optimize the `LIMIT 0` case, we should probably do that separately and go much farther than just removing DISTINCT... Any thoughts on cases where `LIMIT 0` could be an actually useful thing to do (and therefore worth working on)?

**Comment by ranma42** [comment]
I dropped the limit=0 cases; if optimizations are desired for them, they can be way more aggressive and best handled separately (as they basically drop any rows).

**Review by roji — APPROVED** [review]
Thanks @ranma42!

**Pull request merged**